File: cli/lib/worktree.js
/**
 * Git worktree management for parallel spec execution.
 *
 * Creates isolated worktrees at .mint/worktrees/<slug>/ for each parallel spec.
 * Handles creation, env propagation, and cleanup.
 */
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';

/**
 * Create a git worktree for a spec.
 * @param {string} projectRoot - absolute path to the project root
 * @param {string} slug - unique name for the worktree (e.g., "spec-001")
 * @param {object} options
 * @param {string} [options.baseBranch] - branch to base off (default: current HEAD)
 * @returns {{ worktreePath: string, branch: string }}
 */
export function createWorktree(projectRoot, slug, options = {}) {
  const worktreeDir = path.join(projectRoot, '.mint', 'worktrees');
  const worktreePath = path.join(worktreeDir, slug);
  const branch = `mint/${slug}`;

  // Ensure .mint/worktrees/ exists
  fs.mkdirSync(worktreeDir, { recursive: true });

  // Remove stale worktree if it exists
  if (fs.existsSync(worktreePath)) {
    try {
      execSync(`git worktree remove "${worktreePath}" --force`, { cwd: projectRoot, stdio: 'pipe' });
    } catch {
      // Force remove directory if git worktree remove fails
      fs.rmSync(worktreePath, { recursive: true, force: true });
      execSync('git worktree prune', { cwd: projectRoot, stdio: 'pipe' });
    }
  }

  // Delete branch if it exists from a previous run
  try {
    execSync(`git branch -D "${branch}"`, { cwd: projectRoot, stdio: 'pipe' });
  } catch { /* branch doesn't exist, fine */ }

  // Create worktree
  const base = options.baseBranch || 'HEAD';
  execSync(`git worktree add "${worktreePath}" -b "${branch}" ${base}`, {
    cwd: projectRoot,
    stdio: 'pipe',
  });

  // Copy untracked env files into worktree
  copyEnvFiles(projectRoot, worktreePath);

  return { worktreePath, branch };
}

/**
 * Remove a worktree and its branch.
 * @param {string} projectRoot
 * @param {string} slug
 */
export function removeWorktree(projectRoot, slug) {
  const worktreePath = path.join(projectRoot, '.mint', 'worktrees', slug);
  const branch = `mint/${slug}`;

  try {
    execSync(`git worktree remove "${worktreePath}" --force`, { cwd: projectRoot, stdio: 'pipe' });
  } catch {
    if (fs.existsSync(worktreePath)) {
      fs.rmSync(worktreePath, { recursive: true, force: true });
    }
    try { execSync('git worktree prune', { cwd: projectRoot, stdio: 'pipe' }); } catch { /* ignore */ }
  }

  // Clean up branch
  try {
    execSync(`git branch -D "${branch}"`, { cwd: projectRoot, stdio: 'pipe' });
  } catch { /* branch doesn't exist */ }
}

/**
 * List all active mint worktrees.
 * @param {string} projectRoot
 * @returns {Array<{ slug: string, path: string, branch: string }>}
 */
export function listWorktrees(projectRoot) {
  const worktreeDir = path.join(projectRoot, '.mint', 'worktrees');
  if (!fs.existsSync(worktreeDir)) return [];

  try {
    const entries = fs.readdirSync(worktreeDir, { withFileTypes: true });
    return entries
      .filter(e => e.isDirectory())
      .map(e => ({
        slug: e.name,
        path: path.join(worktreeDir, e.name),
        branch: `mint/${e.name}`,
      }));
  } catch {
    return [];
  }
}

/**
 * Remove all mint worktrees (cleanup command).
 * @param {string} projectRoot
 * @returns {number} count of removed worktrees
 */
export function cleanAllWorktrees(projectRoot) {
  const worktrees = listWorktrees(projectRoot);
  let count = 0;
  for (const wt of worktrees) {
    removeWorktree(projectRoot, wt.slug);
    count++;
  }
  // Final prune
  try { execSync('git worktree prune', { cwd: projectRoot, stdio: 'pipe' }); } catch { /* ignore */ }
  return count;
}

/**
 * Copy .env files from project root to worktree.
 * Copies: .env, .env.local, .env.development, .env.development.local
 * @param {string} projectRoot
 * @param {string} worktreePath
 */
function copyEnvFiles(projectRoot, worktreePath) {
  const envPatterns = ['.env', '.env.local', '.env.development', '.env.development.local', '.env.test'];
  for (const pattern of envPatterns) {
    const src = path.join(projectRoot, pattern);
    const dest = path.join(worktreePath, pattern);
    if (fs.existsSync(src)) {
      try { fs.copyFileSync(src, dest); } catch { /* skip if can't copy */ }
    }
  }
}

/**
 * Merge a worktree's changes back to the base branch.
 * Collects all commits from the worktree branch and cherry-picks them.
 * @param {string} projectRoot
 * @param {string} slug
 * @param {string} baseBranch - branch to merge into
 * @returns {{ merged: boolean, commits: number, conflicts: boolean }}
 */
export function mergeWorktree(projectRoot, slug, baseBranch) {
  const branch = `mint/${slug}`;

  try {
    // Get commits unique to the worktree branch
    const log = execSync(`git log ${baseBranch}..${branch} --oneline`, {
      cwd: projectRoot, encoding: 'utf8', stdio: 'pipe',
    }).trim();

    if (!log) return { merged: true, commits: 0, conflicts: false };

    const commitCount = log.split('\n').length;

    // Try merge (no fast-forward to keep history clean)
    try {
      execSync(`git merge ${branch} --no-ff -m "merge: ${slug}"`, {
        cwd: projectRoot, stdio: 'pipe',
      });
      return { merged: true, commits: commitCount, conflicts: false };
    } catch {
      // Merge conflict — abort and report
      try { execSync('git merge --abort', { cwd: projectRoot, stdio: 'pipe' }); } catch { /* ignore */ }
      return { merged: false, commits: commitCount, conflicts: true };
    }
  } catch {
    return { merged: false, commits: 0, conflicts: false };
  }
}

File: cli/internal/worktree/worktree.go
package worktree

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type CreateOptions struct {
	BaseBranch string
}

type Created struct {
	WorktreePath string
	Branch       string
}

type Listed struct {
	Slug   string
	Path   string
	Branch string
}

type CommitResult struct {
	Committed bool
	Reason    string
}

type MergeResult struct {
	Merged    bool
	Commits   int
	Conflicts bool
}

func Create(projectRoot, slug string, options CreateOptions) (Created, error) {
	worktreeDir := filepath.Join(projectRoot, ".mint", "worktrees")
	worktreePath := filepath.Join(worktreeDir, slug)
	branch := "mint/" + slug
	if err := os.MkdirAll(worktreeDir, 0o755); err != nil {
		return Created{}, err
	}
	if _, err := os.Stat(worktreePath); err == nil {
		if err := git(projectRoot, "worktree", "remove", worktreePath, "--force"); err != nil {
			_ = os.RemoveAll(worktreePath)
			_ = git(projectRoot, "worktree", "prune")
		}
	}
	_ = git(projectRoot, "branch", "-D", branch)

	base := options.BaseBranch
	if base == "" {
		base = "HEAD"
	}
	if err := git(projectRoot, "worktree", "add", worktreePath, "-b", branch, base); err != nil {
		return Created{}, err
	}
	copyEnvFiles(projectRoot, worktreePath)
	return Created{WorktreePath: worktreePath, Branch: branch}, nil
}

func Remove(projectRoot, slug string) {
	worktreePath := filepath.Join(projectRoot, ".mint", "worktrees", slug)
	branch := "mint/" + slug
	if err := git(projectRoot, "worktree", "remove", worktreePath, "--force"); err != nil {
		if _, statErr := os.Stat(worktreePath); statErr == nil {
			_ = os.RemoveAll(worktreePath)
		}
		_ = git(projectRoot, "worktree", "prune")
	}
	_ = git(projectRoot, "branch", "-D", branch)
}

func List(projectRoot string) []Listed {
	worktreeDir := filepath.Join(projectRoot, ".mint", "worktrees")
	entries, err := os.ReadDir(worktreeDir)
	if err != nil {
		return nil
	}
	out := make([]Listed, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		slug := entry.Name()
		out = append(out, Listed{
			Slug:   slug,
			Path:   filepath.Join(worktreeDir, slug),
			Branch: "mint/" + slug,
		})
	}
	return out
}

func CleanAll(projectRoot string) int {
	worktrees := List(projectRoot)
	count := 0
	for _, wt := range worktrees {
		Remove(projectRoot, wt.Slug)
		count++
	}
	_ = git(projectRoot, "worktree", "prune")
	return count
}

func CommitChanges(worktreePath, message string) CommitResult {
	dirty, err := gitOutput(worktreePath, "status", "--porcelain", "--", ".", ":!.mint")
	if err != nil {
		return CommitResult{Committed: false, Reason: firstLine(err.Error())}
	}
	if strings.TrimSpace(dirty) == "" {
		return CommitResult{Committed: false, Reason: "clean"}
	}
	if err := git(worktreePath, "add", "-A", "--", ".", ":!.mint"); err != nil {
		return CommitResult{Committed: false, Reason: firstLine(err.Error())}
	}
	if message == "" {
		message = "mint: worker output"
	}
	if err := gitWithInput(worktreePath, message, "commit", "-F", "-"); err != nil {
		return CommitResult{Committed: false, Reason: firstLine(err.Error())}
	}
	return CommitResult{Committed: true}
}

func Merge(projectRoot, slug, baseBranch string) MergeResult {
	branch := "mint/" + slug
	log, err := gitOutput(projectRoot, "log", baseBranch+".."+branch, "--oneline")
	if err != nil {
		return MergeResult{Merged: false, Commits: 0, Conflicts: false}
	}
	log = strings.TrimSpace(log)
	if log == "" {
		return MergeResult{Merged: true, Commits: 0, Conflicts: false}
	}
	commitCount := len(strings.Split(log, "\n"))
	if err := git(projectRoot, "merge", branch, "--no-ff", "-m", "merge: "+slug); err != nil {
		_ = git(projectRoot, "merge", "--abort")
		return MergeResult{Merged: false, Commits: commitCount, Conflicts: true}
	}
	return MergeResult{Merged: true, Commits: commitCount, Conflicts: false}
}

func copyEnvFiles(projectRoot, worktreePath string) {
	for _, name := range []string{".env", ".env.local", ".env.development", ".env.development.local", ".env.test"} {
		src := filepath.Join(projectRoot, name)
		dest := filepath.Join(worktreePath, name)
		content, err := os.ReadFile(src)
		if err != nil {
			continue
		}
		_ = os.WriteFile(dest, content, 0o644)
	}
}

func git(cwd string, args ...string) error {
	_, err := gitCombined(cwd, nil, args...)
	return err
}

func gitOutput(cwd string, args ...string) (string, error) {
	out, err := gitCombined(cwd, nil, args...)
	return string(out), err
}

func gitWithInput(cwd, input string, args ...string) error {
	_, err := gitCombined(cwd, []byte(input), args...)
	return err
}

func gitCombined(cwd string, input []byte, args ...string) ([]byte, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	if input != nil {
		cmd.Stdin = bytes.NewReader(input)
	}
	out, err := cmd.CombinedOutput()
	if err != nil {
		return out, fmt.Errorf("%s: %s", err, strings.TrimSpace(string(out)))
	}
	return out, nil
}

func firstLine(value string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}
	if i := strings.IndexByte(value, '\n'); i >= 0 {
		return value[:i]
	}
	return value
}
